### PR TITLE
Wireup revised secrets model backend

### PR DESCRIPTION
**Pull request by wallyworld**
The secrets hook commands and API have been updated to use the revised secrets data model. This PR updates the apiserver and state layers to match. The secrets rotation backend still needs more work in followup PRs.

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Check secret create, update, get.
Bootstrap and deploy ubuntu
```
$ juju exec --unit ubuntu/0 "secret-add --label foobar --description blah secret!"
secret:cbm3aqfnuod28tu36h50
$ juju secrets
URI                          Revision  Rotate  Backend  Age
secret:cbm3aqfnuod28tu36h50         1  never   juju     5 seconds ago  
$ juju exec --unit ubuntu/0 "secret-get secret:cbm3aqfnuod28tu36h50"
secret!
```

### CODE REVIEW

**Comment by jnsgruk** [comment]
**Drive-by alert :warning: **

```
Age
5 seconds ago
```

Probably needs some rewording :)

**Comment by wallyworld** [comment]
> **Drive-by alert warning **
> 
> ```
> Age
> 5 seconds ago
> ```
> 
> Probably needs some rewording :)

Yeah, things still in a state of flux. Trying to limit the size of each PR. CLI output will get an proper update after backend changes. Everything is still behind a feature flag so we can land stuff incrementally.

**Review by tlm — APPROVED** [review]

File: api/watcher/watcher_test.go [hunk 1]
```diff
@@ -565,6 +572,8 @@ func (s *watcherSuite) setupSecretRotationWatcher(
 			c.Assert(changes, gc.HasLen, 1)
 			c.Assert(changes[0].LastRotateTime.Before(time.Now()), jc.IsTrue)
 			changes[0].LastRotateTime = time.Time{}
+			// TODO(wallyworld) - fix rotation to work with rotate policy
+			change.RotateInterval = 0
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply api/watcher/watcher_test.go:576]: no - the rotation stuff still needs work

File: cmd/juju/secrets/list.go [hunk 2]
```diff
@@ -94,19 +94,21 @@ type secretValueDetails struct {
...
-	Description    string              `json:"description,omitempty" yaml:"description,omitempty"`
-	Owner          string              `json:"owner,omitempty" yaml:"owner,omitempty"`
-	Tags           map[string]string   `json:"tags,omitempty" yaml:"tags,omitempty"`
-	Provider       string              `json:"backend" yaml:"backend"`
-	ProviderID     string              `json:"backend-id,omitempty" yaml:"backend-id,omitempty"`
-	CreateTime     time.Time           `json:"create-time" yaml:"create-time"`
-	UpdateTime     time.Time           `json:"update-time" yaml:"update-time"`
-	Error          string              `json:"error,omitempty" yaml:"error,omitempty"`
-	Value          *secretValueDetails `json:"value,omitempty" yaml:"value,omitempty"`
+	URI            string               `json:"URI" yaml:"URI"`
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply cmd/juju/secrets/list.go:97]: It's like `URL`. The Go SDK also does the same for `URL` etc

File: core/secrets/rotate.go [hunk 3]
```diff
@@ -18,13 +21,55 @@ const (
...
+	switch p {
+	case RotateNever:
+		return nil
+	case RotateHourly:
+		result = lastRotated.Add(time.Hour)
+	case RotateDaily:
+		result = lastRotated.AddDate(0, 0, 1)
+	case RotateWeekly:
+		result = lastRotated.AddDate(0, 0, 7)
+	case RotateHalfMonthly:
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply core/secrets/rotate.go:60]: We will discuss changing the spec, which originally had `BiWeekly` but that's ambiguous.

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 4]
```diff
@@ -90,24 +99,19 @@ func (s *SecretsManagerAPI) UpdateSecrets(args params.UpdateSecretArgs) (params.
...
-	// TODO(wallyworld) - handle label, rotation etc
 	uri, err := coresecrets.ParseURI(arg.URI)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	if uri.ControllerUUID != "" && uri.ControllerUUID != s.controllerUUID {
 		return errors.NotValidf("secret URI with controller UUID %q", uri.ControllerUUID)
 	}
-	if arg.RotatePolicy == nil && arg.Description == nil && arg.Expiry == nil &&
+	if arg.RotatePolicy == nil && arg.Description == nil && arg.ExpireTime == nil &&
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:109]: IMO it's a guard against a programming error - if a struct is passed in with nothing to update, that indicates something ha snot been wired up correctly. the CLI also guards against this

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 5]
```diff
@@ -62,20 +56,35 @@ func (s *SecretsManagerAPI) createSecret(ctx context.Context, arg params.CreateS
...
 	return md.URI.ShortString(), nil
 }
 
+func fromUpsertParams(clock clock.Clock, p params.UpsertSecretArg) secrets.UpsertParams {
+	var nextRotateTime *time.Time
+	if p.RotatePolicy != nil {
+		// TODO(wallyworld) - we need to take into account last rotate time
+		// This approximate will do for now.
+		now := clock.Now()
+		nextRotateTime = p.RotatePolicy.NextRotateTime(&now)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:75]: It's passed by ref to allow the last rotate time to be unknown (ie nil). We could use `time.Time{}` but nil is a common approach to such things.

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/secretsmanager/secrets.go:109]: discussed a bit offline - will leave for now and we can take a view as this progresses

**Pull request merged**